experiments/repo_request.py

```diff
@@ -0,0 +1,108 @@
+import pandas as pd
+import numpy as np
+import os
+import requests
+import json
+
+def load_data(filepath):
+    try:
+        df = pd.read_csv(filepath)
+        print(f'Dataframe shape: {df.shape}')
+        print(f'Dataframe columns: {df.columns.tolist()}')
+        df.info()
+        return df
+
+    except Exception as e:
+        print(f'Error Loading data: {e}')
+    return None
+
+def check_and_clean_data(df):
+    # Checking for null values
+    null_counts = df.isnull().sum()
+    if np.any(null_counts):
+        print("Null values found:")
+        print(null_counts[null_counts > 0])
+    else:
+        print("No null values found.")
+
+    # Checking for duplicates
+    duplicates = df.duplicated().sum()
+    if duplicates:
+        print(f"Number of duplicate rows: {duplicates}")
+    else:
+        print("No duplicate rows found.")
+
+def construct_search_request(repo_path):
+    # constructing the search query
+    search_query = f'test in:path -filename:.txt -filename:.md repo:{repo_path}'
+    payload = {'q': search_query, 'type': 'code'}
+    # The URL for the Search Code endpoint
+    search_url = f'https://api.github.com/search/code'
+    return payload, search_url
+
+def extract_reponame_and_username(repo_path):
+    parts = repo_path.split('/')
+    print(parts)
+    repo_path = '/'.join(parts[-2:])  # This joins the last two parts using '/'
+    return repo_path
+
+def main():
+    data_path = 'df'
+    # Loading the dataframe
+    df = load_data(data_path)
+
+    # Checking the Null and duplicate values
+    check_and_clean_data(df)
+
+    # I will keep the first occurrence of each duplicate row and remove the others:
+    df = df.drop_duplicates(keep='first')
+
+    # I will only consider github.com domain
+    github_df = df[df['repourl'].str.contains('github.com')]
+    df.to_csv('github_df', index = False)
+
+    # Accessing the environmental variable (PAT)
+    pat = os.getenv('MY_PAT')
```

Comment:
If the environment variable is missing:

- what happens in terms of the program's execution? If github rejects the requests then there's no point in continuing to execute if the parameter is missing, if it rate-limits (which is what I think it does) then we might usefully add internal rate-limiting so our code doesn't trigger github's rate-limiting. e.g. we could have a short sleep of say 0.1 seconds before each call to github. 
- should we add the Authorization HTTP header at all if the value is missing?

Reply:
Might be worth adding logging statements with a debug log level to record whether the environment variable was found, and if so, what the value is. Note: the value might be sensitive & valuable if others can obtain it from a logfile there's a potential security vulnerability, so some form of masking might help protect the actual value while also showing enough of a subset of the value for whoever is running the program to see whether the expected value is present or not. As a parallel example, printed receipts from shops often show something like: `paid by credit card 4569 xxxx xxxx 1234`